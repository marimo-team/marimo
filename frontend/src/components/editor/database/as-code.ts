/* Copyright 2024 Marimo. All rights reserved. */
import { assertNever } from "@/utils/assertNever";
import { DatabaseConnectionSchema, type DatabaseConnection } from "./schemas";

export type ConnectionLibrary = "sqlmodel" | "sqlalchemy";
export const ConnectionDisplayNames: Record<ConnectionLibrary, string> = {
  sqlmodel: "SQLModel",
  sqlalchemy: "SQLAlchemy",
};

export function generateDatabaseCode(
  connection: DatabaseConnection,
  orm: ConnectionLibrary,
): string {
  if (!(orm in ConnectionDisplayNames)) {
    throw new Error(`Unsupported library: ${orm}`);
  }

  // Parse the connection to ensure it's valid
  DatabaseConnectionSchema.parse(connection);

  const imports =
    orm === "sqlmodel"
      ? ["from sqlmodel import create_engine", "import os"]
      : ["from sqlalchemy import create_engine", "import os"];

  let code = "";
  switch (connection.type) {
    case "postgres":
      code = `
password = os.environ.get("POSTGRES_PASSWORD", "${connection.password}")
DATABASE_URL = f"postgresql://${connection.username}:{password}@${connection.host}:${connection.port}/${connection.database}"
engine = create_engine(DATABASE_URL${connection.ssl ? ", connect_args={'sslmode': 'require'}" : ""})
`;
      break;

    case "mysql":
      code = `
password = os.environ.get("MYSQL_PASSWORD", "${connection.password}")
DATABASE_URL = f"mysql+pymysql://${connection.username}:{password}@${connection.host}:${connection.port}/${connection.database}"
engine = create_engine(DATABASE_URL${connection.ssl ? ", connect_args={'ssl': {'ssl-mode': 'preferred'}}" : ""})
`;
      break;

    case "sqlite":
      code = `
DATABASE_URL = "sqlite:///${connection.database}"
engine = create_engine(DATABASE_URL)
`;
      break;

    case "snowflake": {
      imports.push(
        "from snowflake.sqlalchemy import URL",
        "import sqlalchemy as sa",
      );

      const params = {
        account: connection.account,
        user: connection.username,
        password: `os.environ.get("SNOWFLAKE_PASSWORD", "${connection.password}")`,
        database: connection.database,
        warehouse: connection.warehouse,
        schema: connection.schema,
        role: connection.role,
      };

      const urlParams = Object.entries(params)
        .filter(([, v]) => v)
        .map(([k, v]) => `        ${k}=${v}`)
        .join(",\n");

      code = `
password = os.environ.get("SNOWFLAKE_PASSWORD", "${connection.password}")
engine = sa.create_engine(
    URL(
${urlParams}
    )
)
`;
      break;
    }

    case "bigquery":
      imports.push("import json");
      code = `
credentials = json.loads("""${connection.credentials_json}""")
engine = create_engine(f"bigquery://${connection.project}/${connection.dataset}", credentials_info=credentials)
`;
      break;

    case "duckdb":
      code = `
engine = create_engine("duckdb:${connection.database || ":memory:"}"${connection.read_only ? ", read_only=True" : ""})
`;
      break;

    default:
      assertNever(connection);
  }

  return `${imports.join("\n")}\n${code.trim()}`;
}
