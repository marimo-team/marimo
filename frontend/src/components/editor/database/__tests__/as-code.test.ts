/* Copyright 2024 Marimo. All rights reserved. */
import { describe, expect, it } from "vitest";
import { type ConnectionLibrary, generateDatabaseCode } from "../as-code";
import type { DatabaseConnection } from "../schemas";

describe("generateDatabaseCode", () => {
  // Test fixtures
  const basePostgres: DatabaseConnection = {
    type: "postgres",
    host: "localhost",
    port: 5432,
    database: "test",
    username: "user",
    password: "pass",
    ssl: true,
  };

  const baseMysql: DatabaseConnection = {
    type: "mysql",
    host: "localhost",
    port: 3306,
    database: "test",
    username: "user",
    password: "pass",
    ssl: true,
  };

  const sqliteConnection: DatabaseConnection = {
    type: "sqlite",
    database: "/path/to/db.sqlite",
  };

  const duckdbConnection: DatabaseConnection = {
    type: "duckdb",
    database: "data.duckdb",
    read_only: true,
  };

  const snowflakeConnection: DatabaseConnection = {
    type: "snowflake",
    account: "account",
    username: "user",
    password: "pass",
    warehouse: "warehouse",
    database: "db",
    schema: "schema",
    role: "role",
  };

  const bigqueryConnection: DatabaseConnection = {
    type: "bigquery",
    project: "my-project",
    dataset: "my_dataset",
    credentials_json: '{"type": "service_account", "project_id": "test"}',
  };

  describe("basic connections", () => {
    it.each([
      ["postgres with SQLModel", basePostgres, "sqlmodel"],
      ["postgres with SQLAlchemy", basePostgres, "sqlalchemy"],
      ["mysql with SQLModel", baseMysql, "sqlmodel"],
      ["mysql with SQLAlchemy", baseMysql, "sqlalchemy"],
      ["sqlite", sqliteConnection, "sqlmodel"],
      ["duckdb", duckdbConnection, "sqlmodel"],
      ["snowflake", snowflakeConnection, "sqlmodel"],
      ["bigquery", bigqueryConnection, "sqlmodel"],
    ])("%s", (name, connection, orm) => {
      expect(
        generateDatabaseCode(connection, orm as ConnectionLibrary),
      ).toMatchSnapshot();
    });
  });

  describe("edge cases", () => {
    const testCases: Array<[string, DatabaseConnection, string]> = [
      [
        "postgres with special chars SQLModel",
        {
          ...basePostgres,
          password: "pass@#$%^&*",
          username: "user-name.special",
          database: "test-db.special",
        },
        "sqlmodel",
      ],
      [
        "postgres with special chars SQLAlchemy",
        {
          ...basePostgres,
          password: "pass@#$%^&*",
          username: "user-name.special",
          database: "test-db.special",
        },
        "sqlalchemy",
      ],
      [
        "snowflake with minimal config SQLModel",
        {
          type: "snowflake",
          account: "account",
          username: "user",
          password: "pass",
          database: "db",
          warehouse: "",
          schema: "",
          role: "",
        },
        "sqlmodel",
      ],
      [
        "postgres with unicode",
        {
          ...basePostgres,
          database: "测试数据库",
          username: "用户",
          password: "密码",
        },
        "sqlmodel",
      ],
      [
        "bigquery with long credentials",
        {
          ...bigqueryConnection,
          credentials_json: "x".repeat(10),
        },
        "sqlmodel",
      ],
      [
        "sqlite with empty path",
        {
          type: "sqlite",
          // sqlite allows empty path
          database: "",
        },
        "sqlmodel",
      ],
      [
        "postgres with IPv6",
        {
          ...basePostgres,
          host: "2001:0db8:85a3:0000:0000:8a2e:0370:7334",
        },
        "sqlmodel",
      ],
      [
        "postgres with non-standard port",
        {
          ...basePostgres,
          port: 54_321,
        },
        "sqlmodel",
      ],
      [
        "mysql with max port",
        {
          ...baseMysql,
          port: 65_535,
        },
        "sqlmodel",
      ],
      [
        "postgres with URL-encoded characters",
        {
          ...basePostgres,
          database: "test%20db",
          username: "user%20name",
          password: "pass%20word",
        },
        "sqlmodel",
      ],
      [
        "mysql with extremely long database name",
        {
          ...baseMysql,
          database: "x".repeat(64),
        },
        "sqlmodel",
      ],
      [
        "snowflake with all optional fields filled",
        {
          type: "snowflake",
          account: "org-account",
          username: "user",
          password: "pass",
          database: "db",
          warehouse: "compute_wh",
          schema: "public",
          role: "accountadmin",
        },
        "sqlmodel",
      ],
      [
        "duckdb with relative path",
        {
          type: "duckdb",
          database: "./relative/path/db.duckdb",
          read_only: false,
        },
        "sqlmodel",
      ],
      [
        "postgres with domain socket",
        {
          ...basePostgres,
          host: "/var/run/postgresql",
          // @ts-expect-error - Testing invalid input
          port: undefined,
        },
        "sqlmodel",
      ],
    ];

    it.each(testCases)("%s", (name, connection, orm) => {
      expect(
        generateDatabaseCode(connection, orm as ConnectionLibrary),
      ).toMatchSnapshot();
    });
  });

  describe("security cases", () => {
    it.each([
      [
        "postgres with empty password",
        {
          ...basePostgres,
          password: "",
        },
      ],
      [
        "mysql with very long password",
        {
          ...baseMysql,
          password: "x".repeat(10),
        },
      ],
      [
        "postgres with SQL injection attempt in database name",
        {
          ...basePostgres,
          database: "db'; DROP TABLE users;--",
        },
      ],
      [
        "snowflake with sensitive info in account",
        {
          ...snowflakeConnection,
          account: "account-with-password123",
        },
      ],
      [
        "bigquery with malformed JSON",
        {
          ...bigqueryConnection,
          credentials_json: '{"type": "service_account", "project_id": "test"',
        },
      ],
    ])("%s", (name, connection) => {
      expect(generateDatabaseCode(connection, "sqlmodel")).toMatchSnapshot();
      expect(generateDatabaseCode(connection, "sqlalchemy")).toMatchSnapshot();
    });
  });

  describe("invalid cases", () => {
    it.each([
      [
        "throws for unsupported ORMs",
        () =>
          // @ts-expect-error - Testing invalid input
          generateDatabaseCode(basePostgres, "polars"),
      ],
      [
        "throws for invalid port",
        () => generateDatabaseCode({ ...basePostgres, port: -1 }, "sqlmodel"),
      ],
      [
        "throws for invalid host",
        () => generateDatabaseCode({ ...basePostgres, host: "" }, "sqlmodel"),
      ],
      [
        "throws for port out of range",
        () =>
          generateDatabaseCode({ ...basePostgres, port: 65_536 }, "sqlmodel"),
      ],
      [
        "throws for invalid snowflake account",
        () =>
          generateDatabaseCode(
            { ...snowflakeConnection, account: "" },
            "sqlmodel",
          ),
      ],
      [
        "throws for invalid bigquery project",
        () =>
          generateDatabaseCode(
            { ...bigqueryConnection, project: "" },
            "sqlmodel",
          ),
      ],
    ])("%s", (name, fn) => {
      expect(fn).toThrow();
    });
  });
});
