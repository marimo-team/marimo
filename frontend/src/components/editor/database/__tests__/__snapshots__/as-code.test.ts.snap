// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateDatabaseCode > basic connections > bigquery 1`] = `
"from sqlmodel import create_engine
import os
import json
credentials = json.loads("""{"type": "service_account", "project_id": "test"}""")
engine = create_engine(f"bigquery://my-project/my_dataset", credentials_info=credentials)"
`;

exports[`generateDatabaseCode > basic connections > duckdb 1`] = `
"from sqlmodel import create_engine
import os
engine = create_engine("duckdb:data.duckdb", read_only=True)"
`;

exports[`generateDatabaseCode > basic connections > mysql with SQLAlchemy 1`] = `
"from sqlalchemy import create_engine
import os
password = os.environ.get("MYSQL_PASSWORD", "pass")
DATABASE_URL = f"mysql+pymysql://user:{password}@localhost:3306/test"
engine = create_engine(DATABASE_URL, connect_args={'ssl': {'ssl-mode': 'preferred'}})"
`;

exports[`generateDatabaseCode > basic connections > mysql with SQLModel 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("MYSQL_PASSWORD", "pass")
DATABASE_URL = f"mysql+pymysql://user:{password}@localhost:3306/test"
engine = create_engine(DATABASE_URL, connect_args={'ssl': {'ssl-mode': 'preferred'}})"
`;

exports[`generateDatabaseCode > basic connections > postgres with SQLAlchemy 1`] = `
"from sqlalchemy import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass")
DATABASE_URL = f"postgresql://user:{password}@localhost:5432/test"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > basic connections > postgres with SQLModel 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass")
DATABASE_URL = f"postgresql://user:{password}@localhost:5432/test"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > basic connections > snowflake 1`] = `
"from sqlmodel import create_engine
import os
from snowflake.sqlalchemy import URL
import sqlalchemy as sa
password = os.environ.get("SNOWFLAKE_PASSWORD", "pass")
engine = sa.create_engine(
    URL(
        account=account,
        user=user,
        password=os.environ.get("SNOWFLAKE_PASSWORD", "pass"),
        database=db,
        warehouse=warehouse,
        schema=schema,
        role=role
    )
)"
`;

exports[`generateDatabaseCode > basic connections > sqlite 1`] = `
"from sqlmodel import create_engine
import os
DATABASE_URL = "sqlite:////path/to/db.sqlite"
engine = create_engine(DATABASE_URL)"
`;

exports[`generateDatabaseCode > edge cases > bigquery with long credentials 1`] = `
"from sqlmodel import create_engine
import os
import json
credentials = json.loads("""xxxxxxxxxx""")
engine = create_engine(f"bigquery://my-project/my_dataset", credentials_info=credentials)"
`;

exports[`generateDatabaseCode > edge cases > duckdb with relative path 1`] = `
"from sqlmodel import create_engine
import os
engine = create_engine("duckdb:./relative/path/db.duckdb")"
`;

exports[`generateDatabaseCode > edge cases > mysql with extremely long database name 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("MYSQL_PASSWORD", "pass")
DATABASE_URL = f"mysql+pymysql://user:{password}@localhost:3306/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
engine = create_engine(DATABASE_URL, connect_args={'ssl': {'ssl-mode': 'preferred'}})"
`;

exports[`generateDatabaseCode > edge cases > mysql with max port 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("MYSQL_PASSWORD", "pass")
DATABASE_URL = f"mysql+pymysql://user:{password}@localhost:65535/test"
engine = create_engine(DATABASE_URL, connect_args={'ssl': {'ssl-mode': 'preferred'}})"
`;

exports[`generateDatabaseCode > edge cases > postgres with IPv6 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass")
DATABASE_URL = f"postgresql://user:{password}@2001:0db8:85a3:0000:0000:8a2e:0370:7334:5432/test"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > edge cases > postgres with URL-encoded characters 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass%20word")
DATABASE_URL = f"postgresql://user%20name:{password}@localhost:5432/test%20db"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > edge cases > postgres with domain socket 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass")
DATABASE_URL = f"postgresql://user:{password}@/var/run/postgresql:undefined/test"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > edge cases > postgres with non-standard port 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass")
DATABASE_URL = f"postgresql://user:{password}@localhost:54321/test"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > edge cases > postgres with special chars SQLAlchemy 1`] = `
"from sqlalchemy import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass@#$%^&*")
DATABASE_URL = f"postgresql://user-name.special:{password}@localhost:5432/test-db.special"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > edge cases > postgres with special chars SQLModel 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass@#$%^&*")
DATABASE_URL = f"postgresql://user-name.special:{password}@localhost:5432/test-db.special"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > edge cases > postgres with unicode 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "密码")
DATABASE_URL = f"postgresql://用户:{password}@localhost:5432/测试数据库"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > edge cases > snowflake with all optional fields filled 1`] = `
"from sqlmodel import create_engine
import os
from snowflake.sqlalchemy import URL
import sqlalchemy as sa
password = os.environ.get("SNOWFLAKE_PASSWORD", "pass")
engine = sa.create_engine(
    URL(
        account=org-account,
        user=user,
        password=os.environ.get("SNOWFLAKE_PASSWORD", "pass"),
        database=db,
        warehouse=compute_wh,
        schema=public,
        role=accountadmin
    )
)"
`;

exports[`generateDatabaseCode > edge cases > snowflake with minimal config SQLModel 1`] = `
"from sqlmodel import create_engine
import os
from snowflake.sqlalchemy import URL
import sqlalchemy as sa
password = os.environ.get("SNOWFLAKE_PASSWORD", "pass")
engine = sa.create_engine(
    URL(
        account=account,
        user=user,
        password=os.environ.get("SNOWFLAKE_PASSWORD", "pass"),
        database=db
    )
)"
`;

exports[`generateDatabaseCode > edge cases > sqlite with empty path 1`] = `
"from sqlmodel import create_engine
import os
DATABASE_URL = "sqlite:///"
engine = create_engine(DATABASE_URL)"
`;

exports[`generateDatabaseCode > security cases > bigquery with malformed JSON 1`] = `
"from sqlmodel import create_engine
import os
import json
credentials = json.loads("""{"type": "service_account", "project_id": "test"""")
engine = create_engine(f"bigquery://my-project/my_dataset", credentials_info=credentials)"
`;

exports[`generateDatabaseCode > security cases > bigquery with malformed JSON 2`] = `
"from sqlalchemy import create_engine
import os
import json
credentials = json.loads("""{"type": "service_account", "project_id": "test"""")
engine = create_engine(f"bigquery://my-project/my_dataset", credentials_info=credentials)"
`;

exports[`generateDatabaseCode > security cases > mysql with very long password 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("MYSQL_PASSWORD", "xxxxxxxxxx")
DATABASE_URL = f"mysql+pymysql://user:{password}@localhost:3306/test"
engine = create_engine(DATABASE_URL, connect_args={'ssl': {'ssl-mode': 'preferred'}})"
`;

exports[`generateDatabaseCode > security cases > mysql with very long password 2`] = `
"from sqlalchemy import create_engine
import os
password = os.environ.get("MYSQL_PASSWORD", "xxxxxxxxxx")
DATABASE_URL = f"mysql+pymysql://user:{password}@localhost:3306/test"
engine = create_engine(DATABASE_URL, connect_args={'ssl': {'ssl-mode': 'preferred'}})"
`;

exports[`generateDatabaseCode > security cases > postgres with SQL injection attempt in database name 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass")
DATABASE_URL = f"postgresql://user:{password}@localhost:5432/db'; DROP TABLE users;--"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > security cases > postgres with SQL injection attempt in database name 2`] = `
"from sqlalchemy import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "pass")
DATABASE_URL = f"postgresql://user:{password}@localhost:5432/db'; DROP TABLE users;--"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > security cases > postgres with empty password 1`] = `
"from sqlmodel import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "")
DATABASE_URL = f"postgresql://user:{password}@localhost:5432/test"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > security cases > postgres with empty password 2`] = `
"from sqlalchemy import create_engine
import os
password = os.environ.get("POSTGRES_PASSWORD", "")
DATABASE_URL = f"postgresql://user:{password}@localhost:5432/test"
engine = create_engine(DATABASE_URL, connect_args={'sslmode': 'require'})"
`;

exports[`generateDatabaseCode > security cases > snowflake with sensitive info in account 1`] = `
"from sqlmodel import create_engine
import os
from snowflake.sqlalchemy import URL
import sqlalchemy as sa
password = os.environ.get("SNOWFLAKE_PASSWORD", "pass")
engine = sa.create_engine(
    URL(
        account=account-with-password123,
        user=user,
        password=os.environ.get("SNOWFLAKE_PASSWORD", "pass"),
        database=db,
        warehouse=warehouse,
        schema=schema,
        role=role
    )
)"
`;

exports[`generateDatabaseCode > security cases > snowflake with sensitive info in account 2`] = `
"from sqlalchemy import create_engine
import os
from snowflake.sqlalchemy import URL
import sqlalchemy as sa
password = os.environ.get("SNOWFLAKE_PASSWORD", "pass")
engine = sa.create_engine(
    URL(
        account=account-with-password123,
        user=user,
        password=os.environ.get("SNOWFLAKE_PASSWORD", "pass"),
        database=db,
        warehouse=warehouse,
        schema=schema,
        role=role
    )
)"
`;
